sdk/spring/azure-spring-boot/src/main/java/com/microsoft/azure/keyvault/spring/KeyVaultOperation.java

```diff
@@ -0,0 +1,151 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.microsoft.azure.keyvault.spring;
+
+import com.azure.core.http.rest.PagedIterable;
+import com.azure.security.keyvault.secrets.SecretClient;
+import com.azure.security.keyvault.secrets.models.KeyVaultSecret;
+import com.azure.security.keyvault.secrets.models.SecretProperties;
+import org.springframework.lang.NonNull;
+import org.springframework.util.Assert;
+import org.springframework.util.StringUtils;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Locale;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+
+/**
+ * Encapsulate key vault secret client in this class to provide a delegate of key vault operations.
+ */
+public class KeyVaultOperation {
+    private final long cacheRefreshIntervalInMs;
+    private final List<String> secretKeys;
+
+    private final Object refreshLock = new Object();
+    private final SecretClient keyVaultClient;
+    private final String vaultUri;
+
+    private ArrayList<String> propertyNames = new ArrayList<>();
+    private String[] propertyNamesArr;
+
+    private final AtomicLong lastUpdateTime = new AtomicLong();
+    private final ReadWriteLock rwLock = new ReentrantReadWriteLock();
+
+    public KeyVaultOperation(final SecretClient keyVaultClient,
+                             String vaultUri,
+                             final long refreshInterval,
+                             final List<String> secretKeys) {
+        this.cacheRefreshIntervalInMs = refreshInterval;
+        this.secretKeys = secretKeys;
+        this.keyVaultClient = keyVaultClient;
+        // TODO(pan): need to validate why last '/' need to be truncated.
+        this.vaultUri = StringUtils.trimTrailingCharacter(vaultUri.trim(), '/');
+        fillSecretsList();
+    }
+
+    public String[] list() {
+        try {
+            this.rwLock.readLock().lock();
+            return propertyNamesArr;
+        } finally {
+            this.rwLock.readLock().unlock();
+        }
```

Comment:
Do you really need a lock? Can declaring `propertyNamesArr` as volatile be sufficient?

Reply:
Yes, the write method `fillSecretsList` could be called from multiple threads simultaneously when refreshing `propertyNames`.